go-controller/pkg/ovn/address_set_sync.go

```diff
@@ -0,0 +1,358 @@
+		// oldName is just namespace name
+		dbIndex = &addressset.DbIndex{
+			OwnerObjectType: libovsdbops.NamespaceOwnerType,
+			OwnerName:       oldName,
+		}
+	}
+	// dbIndex is set, find referencing objects
+	qosPred := func(qos *nbdb.QoS) bool {
+		return strings.Contains(qos.Match, "$"+oldHash)
+	}
+	qoses, err = libovsdbops.FindQoSesWithPredicate(nbClient, qosPred)
+	if err != nil {
+		err = fmt.Errorf("failed to find qoses for address set %s: %v", oldHash, err)
+		return
+	}
+	lrpPred := func(lrp *nbdb.LogicalRouterPolicy) bool {
+		return strings.Contains(lrp.Match, "$"+oldHash)
+	}
+	lrps, err = libovsdbops.FindLogicalRouterPoliciesWithPredicate(nbClient, lrpPred)
+	if err != nil {
+		err = fmt.Errorf("failed to find lrps for address set %s: %v", oldHash, err)
+		return
+	}
+	return
+}
+
+func (syncer *addressSetsSyncer) getDeleteASOps(acls []*nbdb.ACL, qoses []*nbdb.QoS, lrps []*nbdb.LogicalRouterPolicy,
+	as *nbdb.AddressSet) (ops []libovsdb.Operation, err error) {
+	if len(acls) > 0 {
+		// delete acls from all port groups and switches, since we don't know where exactly it is referenced
+		ops, err = libovsdbops.DeleteACLsFromPortGroupWithPredicateOps(syncer.nbClient, ops,
+			func(item *nbdb.PortGroup) bool { return true }, acls...)
+		if err != nil {
+			err = fmt.Errorf("failed to get acl delete from port group ops: %v", err)
+			return
+		}
+		ops, err = libovsdbops.RemoveACLsFromLogicalSwitchesWithPredicateOps(syncer.nbClient, ops,
+			func(item *nbdb.LogicalSwitch) bool { return true }, acls...)
+		if err != nil {
+			err = fmt.Errorf("failed to get acl delete from logical switch ops: %v", err)
+			return
+		}
+	}
+	if len(qoses) > 0 {
+		ops, err = libovsdbops.RemoveQoSsFromLogicalSwitchesWithPredicateOps(syncer.nbClient, ops,
+			func(item *nbdb.LogicalSwitch) bool { return true }, qoses...)
+		if err != nil {
+			err = fmt.Errorf("failed to get qos delete from logical switch ops: %v", err)
+			return
+		}
+		ops, err = libovsdbops.DeleteQoSesOps(syncer.nbClient, ops, qoses...)
+		if err != nil {
+			err = fmt.Errorf("failed to get qos delete ops: %v", err)
+			return
+		}
+	}
+	if len(lrps) > 0 {
+		ops, err = libovsdbops.DeleteLogicalRouterPoliciesOps(syncer.nbClient, ops, ovntypes.OVNClusterRouter, lrps...)
+		if err != nil {
+			err = fmt.Errorf("failed to get logical router policy delete ops: %v", err)
+			return
+		}
+	}
+	ops, err = libovsdbops.DeleteAddressSetsOps(syncer.nbClient, ops, as)
+	return
+}
+
+// updateOpsForAS adds db ops to update address set and objects that reference it
+func (syncer *addressSetsSyncer) updateOpsForAS(as *nbdb.AddressSet) error {
+	oldName, ipSuffix := truncateSuffixFromAddressSet(as.ExternalIDs["name"])
+	// oldName may be empty if as doesn't have ExternalID set
+	acls, qoses, lrps, dbIndex, err := GetReferencingObjsAndNewDbIndex(syncer.nbClient, as.Name, oldName)
+	if err != nil {
+		return fmt.Errorf("failed to get new dbIndex for address set %s: %v", oldName, err)
```

Comment:
so if this fails we are going to crash? Should we just remove the address set and consider it stale instead along with any acls, qoses, lrps that reference it?

Reply:
What do you mean by "remove the address set and consider it stale"?
`GetReferencingObjsAndNewDbIndex` returns error if search for referencing objects fails, so we can't delete the address set, because there can be acls/qoses/etc referencing this address set. The best we can do is just leave it in the db and retry later.

Reply:
What I was concerned about was returning an error here because we found something we didnt expect, and I was thinking we should just remove that thing along with its references. But the only time err is set here looks like if we cannot transact correctly with libovsdb, like:

```
	qoses, err = libovsdbops.FindQoSesWithPredicate(nbClient, qosPred)
	if err != nil {
		err = fmt.Errorf("failed to find qoses for address set %s: %v", oldHash, err)
		return
	}
```

In that case it is a transient error with libovsdb and retrying should be OK. Agree?

Reply:
That would be nice, but we don't have any retry mechanism for the initial sync functions, so if the same happens e.g, for netpol sync (all the other syncs work the same afaik) https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/policy.go#L354, `(oc *DefaultNetworkController) Run` will return an error, and the pod will be restarted. So I am not sure if implementing a retry for this specific address set sync makes sense, wdyt?